File: backend_node/models/Progress.js
const mongoose = require('mongoose');

const mistakeSchema = new mongoose.Schema({
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    lesson: { type: mongoose.Schema.Types.ObjectId, ref: 'Lesson' },
    error_type: { type: String, required: true },           // e.g., 'madd', 'ghunna', 'qalqala'
    rule_category: { type: String, default: '' },            // e.g., 'المدود', 'أحكام النون الساكنة'
    rule_name_ar: { type: String, default: '' },             // e.g., 'المد', 'الغنة'
    surah_number: { type: Number },
    ayah_number: { type: Number },
    ayah_text: { type: String },
    char_index: { type: Number },
    is_corrected: { type: Boolean, default: false },
    corrected_at: { type: Date },
    feedback: { type: String },
    audio_url: { type: String },
    occurrence_count: { type: Number, default: 1 }
}, { timestamps: true });

const progressSchema = new mongoose.Schema({
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    lesson: { type: mongoose.Schema.Types.ObjectId, ref: 'Lesson', required: true },
    status: {
        type: String,
        enum: ['not-started', 'in-progress', 'completed'],
        default: 'not-started'
    },
    score: { type: Number, default: 0 },
    last_accessed: { type: Date, default: Date.now }
}, { timestamps: true });

progressSchema.index({ user: 1, lesson: 1 }, { unique: true });

const Progress = mongoose.model('Progress', progressSchema);
const Mistake = mongoose.model('Mistake', mistakeSchema);

module.exports = { Progress, Mistake };

File: backend/app/models/progress.py
"""
Progress & Mistake Pydantic models.
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum


class ProgressStatus(str, Enum):
    NOT_STARTED = "not-started"
    IN_PROGRESS = "in-progress"
    COMPLETED = "completed"


# ── Request Schemas ──

class ProgressUpdate(BaseModel):
    lesson_id: str
    status: ProgressStatus
    score: int = 0
    theoretical_score: int = 0
    practical_tests: dict = Field(default_factory=dict) # {"test_id": "passed" / "failed"}
    video_watched: bool = False


class MistakeLog(BaseModel):
    error_type: str
    lesson_id: Optional[str] = None
    surah_number: Optional[int] = None
    ayah_number: Optional[int] = None
    ayah_text: Optional[str] = ""
    char_index: Optional[int] = None
    feedback: Optional[str] = None
    audio_url: Optional[str] = None


class MarkCorrected(BaseModel):
    error_type: str
    surah_number: int
    ayah_number: int


# ── Error Type Mapping (server-side) ──

ERROR_TYPE_MAP = {
    "madd":      {"category": "أحكام المدود", "name": "خطأ في المد"},
    "ghunna":    {"category": "الغنة وأحكام النون والميم", "name": "خطأ في الغنة"},
    "qalqala":   {"category": "القلقلة", "name": "خطأ في القلقلة"},
    "vowel":     {"category": "الحركات والتشكيل", "name": "خطأ في الحركات"},
    "tafkheem":  {"category": "التفخيم والترقيق", "name": "خطأ في التفخيم أو الترقيق"},
    "hams_jahr": {"category": "الهمس والجهر", "name": "خطأ في الهمس أو الجهر"},
    "shidda":    {"category": "الشدة والرخاوة", "name": "خطأ في الشدة أو الرخاوة"},
    "safeer":    {"category": "الصفير", "name": "خطأ في الصفير"},
    "istitala":  {"category": "الاستطالة", "name": "خطأ في الاستطالة"},
    "sifat":     {"category": "صفات الحروف", "name": "خطأ في صفة الحرف"},
    "phoneme":   {"category": "مخارج الحروف", "name": "خطأ في مخرج الحرف"},
    "deletion":  {"category": "أخطاء النطق", "name": "حذف حرف أو صوت"},
    "insertion": {"category": "أخطاء النطق", "name": "إضافة صوت زائد"},
}
